### PR TITLE
[spark][core] Support more expressions in V2 filter pushdown

### PR DESCRIPTION
**Pull request by JingsongLi**
### Purpose

Expand Spark DataSource V2 filter conversion with direct-column support for:

- `<>` / not-equal predicates
- `CHAR_LENGTH`
- `YEAR`, `MONTH`, `DAY`, `HOUR`, `MINUTE`, and `SECOND` extraction for `DATE` and timestamp-without-time-zone fields
- `NOT LIKE` pushdown for Spark V2 `NOT(STARTS_WITH)`, `NOT(ENDS_WITH)`, and `NOT(CONTAINS)` predicates, preserving SQL null semantics
- Safe fallback behavior for the new negative string predicates in index and format visitors

When legacy timestamp mapping exposes a Paimon timestamp-without-time-zone as Spark `TIMESTAMP`, date/time extraction remains in Spark to preserve session-time-zone semantics.

### Tests

- `paimon-common` predicate and JSON serde tests: 141 tests
- `paimon-format`: 57 ORC and Parquet filter tests
- Spark 3.5 `SparkV2FilterConverterTest`: 37 tests
- Spark 3.3 compatibility `SparkV2FilterConverterTest`: 37 tests
- Spark 3.5 `PaimonPushDownTest`: 22 tests
- `SparkFilterConverterTest`: 7 tests on Spark 3 / Scala 2.12 and Spark 4 / Scala 2.13
- Spark 4 / Scala 2.13 affected modules compile successfully
- Non-fast Spotless, Checkstyle, and compile verification for affected modules

### CODE REVIEW

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/util/SparkExpressionConverter.scala [hunk 1]
```diff
@@ -77,12 +87,26 @@ object SparkExpressionConverter {
...
+        e match {
+          case extract: Extract =>
+            extract.source() match {
+              case n: NamedReference =>
+                val fieldRef = toPaimonFieldRef(n, rowType)
+                extract.field() match {
+                  case EXTRACT_YEAR => YearTransform.tryCreate(fieldRef)
+                  case EXTRACT_MONTH => MonthTransform.tryCreate(fieldRef)
+                  case EXTRACT_DAY => DayTransform.tryCreate(fieldRef)
+                  case EXTRACT_HOUR => HourTransform.tryCreate(fieldRef)
```
Comment (leaves12138) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/util/SparkExpressionConverter.scala]: [P1] Keep legacy-mapped timestamp extraction in Spark

Checking only the Paimon field type is insufficient here. With `spark.paimon.legacy-timestamp-mapping.enabled=true`, a Paimon `TIMESTAMP_WITHOUT_TIME_ZONE` is exposed as Spark `TIMESTAMP`, so `hour(ts)` is evaluated in the Spark session time zone. `HourTransform` instead extracts the stored local timestamp directly. These differ when the Spark session and JVM time zones differ. For a timestamp partition column, Spark considers this predicate fully handled and removes its residual filter, while Paimon prunes the matching partition.

Reproduced on this head with Spark 3.5.8, JVM time zone UTC, and both ORC and Parquet:

```sql
SET spark.paimon.legacy-timestamp-mapping.enabled=true;
SET spark.sql.session.timeZone=America/Los_Angeles;
CREATE TABLE legacy_extract (id INT, ts TIMESTAMP)
USING paimon PARTITIONED BY (ts);
INSERT INTO legacy_extract VALUES (1, timestamp'2025-01-15 01:02:03');
SELECT hour(ts) FROM legacy_extract;
-- Returns 1.
SELECT id FROM legacy_extract WHERE hour(ts) = 1;
-- Expected [1], actual empty result.
```

The physical plan contains `PartitionFilters: [And([IsNotNull(ts), Equal(HOUR(ts), 1)])]` with no Spark residual filter. Please decline timestamp extraction when `treatPaimonTimestampTypeAsSparkTimestampType()` is enabled, unless the transform explicitly reproduces Spark's time-zone semantics. A temporary local guard for this case fixed both reproductions; all 42 converter/boundary tests passed with that guard. Please also add a partitioned-table regression with different JVM and session time zones; the existing LTZ fallback test only covers the default type mapping.

**Review by leaves12138 — CHANGES_REQUESTED** [review]
Reviewed 6d14883023e642a9d3fd24eff42d6909bcd3bb26. Requesting changes for one reproducible wrong-result issue: date/time extraction can incorrectly prune timestamp partitions under legacy Spark timestamp mapping when JVM and session time zones differ. Details and a minimal reproduction are in the inline comment.

Validation on the unmodified production code:
- 298 targeted common predicate/transform/JSON tests passed.
- 57 ORC/Parquet filter tests passed.
- 36 Spark 3.5 V2 filter converter tests passed.
- 36 Spark 3.3 V2 filter converter compatibility tests passed.
- Six additional Spark 3.5 boundary tests: four passed (ordinary-table legacy timestamps, negative string predicates with NULL/OR, and Unicode character length); two failed with actual missing rows in timestamp-partitioned tables, independently for ORC and Parquet.

A temporary local fallback guard made all 42 Spark 3.5 converter/boundary tests pass. That experimental production-code change has been reverted; no fix was pushed. The new negative string predicates' null handling and conservative index/format fallbacks did not reveal another blocker in the exercised paths.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/util/SparkExpressionConverter.scala (hunk 1 above)
Comment (JingsongLi) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/util/SparkExpressionConverter.scala]: Fixed in 345260b08a. `Extract` conversion now declines Paimon `TIMESTAMP_WITHOUT_TIME_ZONE` when legacy timestamp mapping exposes it as Spark `TIMESTAMP`, so Spark retains the time-zone-sensitive predicate. DATE extraction and normal TIMESTAMP_NTZ extraction remain pushable. Added a partitioned-table regression with JVM UTC and Spark session America/Los_Angeles; the Spark 3.5 converter suite now passes all 37 tests.

**Review by leaves12138 — APPROVED** [review]
Re-reviewed 345260b08a4f270fcbad805ebb469e2f897ca67f, including the delta from my previous review. The legacy timestamp partition-pruning issue is fixed, and I found no remaining blocking issue in this update.

The new guard correctly keeps timestamp extraction in Spark when legacy timestamp mapping is enabled, without disabling DATE or normal timestamp-without-time-zone extraction. I reran the original end-to-end reproductions with JVM time zone UTC and Spark session time zone America/Los_Angeles: both ORC and Parquet timestamp-partitioned tables now return the expected row for `hour(ts) = 1`.

Validation on this head:
- Spark 3.5: all 43 Scala tests passed (37 V2 converter tests plus 6 independent boundary cases, including the original missing-row reproductions).
- Spark 3.3: all 37 Scala V2 converter compatibility tests passed.
- All 7 Java `SparkFilterConverterTest` tests passed in each run, including the updated supported/unsupported predicate assertions.

The broader GitHub CI is still in progress; this approval reflects code review and the targeted checks above, not a claim that every CI lane has completed. No production-code workaround was applied for this validation, and no code was pushed.
